kubernetes/updater/pkg/controller/deployment.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"github.com/docker/distribution/reference"
+	"github.com/gravitational/trace"
+	"golang.org/x/mod/semver"
+	appsv1 "k8s.io/api/apps/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// DeploymentVersionUpdater Reconciles a deployment by changing its img
+type DeploymentVersionUpdater struct {
+	VersionUpdater
+	kclient.Client
+	Scheme *runtime.Scheme
+}
+
+func (r *DeploymentVersionUpdater) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrllog.FromContext(ctx).WithValues("namespacedname", req.NamespacedName)
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	// Get the object we are reconciling
+	var obj appsv1.Deployment
+	if err := r.Get(ctx, req.NamespacedName, &obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			// log.Info("not found")
+			return ctrl.Result{}, nil
+		}
+		// log.Error(err, "failed to get resource")
+		return ctrl.Result{}, trace.Wrap(err)
+	}
+
+	// Get the current and past version
+	currentVersion, err := getDeploymentVersion(&obj)
+	if err != nil {
+		switch trace.Unwrap(err).(type) {
+		case *trace.BadParameterError:
+			log.Info("Teleport container found, but failed to get version from the img tag. Will continue and do a version update.")
+		default:
+			return requeueLater, trace.Wrap(err)
+		}
+	}
+
+	image, err := r.Do(ctx, &obj, currentVersion)
+	var (
+		noNewVersionErr *NoNewVersionError
+		maintenanceErr  *MaintenanceNotTriggeredError
+	)
+	switch {
+	case errors.As(err, &noNewVersionErr):
+		// Error contains the detected versions
+		log.Info("Deployment is already up-to-date, not updating.", "err", err)
+		return requeueLater, nil
+	case errors.As(err, &maintenanceErr):
+		// Not logging the error because it provides no other information than its type.
+		log.Info("No maintenance triggered, not updating.")
```

Comment:
Should we log the current version and the upstream version in case we have some bug and it's easier to track?

Reply:
If the maintenance is not triggered we don't retrieve the new version, we bail out before. I'll log the current version.

If the deployment is not updated because the current and new version are identical, it will be logged through the `noNewVersion` logging happening above.

Reply:
This made me realize the image tag was not logged in validator errors, and the image version was not logged after a successful update.